File: currency_favorites.php
# currency_favorites.php
#!/usr/bin/env php
<?php

define('FAVORITES_FILE', 'favorites.json');
define('API_URL', 'https://api.frankfurter.app/latest?from=');

function loadFavorites() {
    if (file_exists(FAVORITES_FILE)) {
        $data = json_decode(file_get_contents(FAVORITES_FILE), true);
        return $data['pairs'] ?? [['USD', 'EUR'], ['USD', 'JPY']];
    }
    return [['USD', 'EUR'], ['USD', 'JPY']];
}

function saveFavorites($pairs) {
    file_put_contents(FAVORITES_FILE, json_encode(['pairs' => $pairs], JSON_PRETTY_PRINT));
}

function getRates($base) {
    $url = API_URL . $base;
    $resp = file_get_contents($url);
    if ($resp === false) {
        return [];
    }
    $data = json_decode($resp, true);
    return $data['rates'] ?? [];
}

function listRates($pairs) {
    if (empty($pairs)) {
        echo "No favorite pairs.\n";
        return;
    }
    $grouped = [];
    foreach ($pairs as $pair) {
        $base = $pair[0];
        $target = $pair[1];
        if (!isset($grouped[$base])) $grouped[$base] = [];
        $grouped[$base][] = $target;
    }
    echo "\n💱 Favorite Currency Pairs\n\n";
    foreach ($grouped as $base => $targets) {
        $rates = getRates($base);
        foreach ($targets as $target) {
            if (isset($rates[$target])) {
                printf("%s/%s: %.4f\n", $base, $target, $rates[$target]);
            } else {
                echo "$base/$target: N/A\n";
            }
        }
    }
    echo "\nUpdated: " . date('Y-m-d H:i:s') . "\n";
}

if ($argc < 2) {
    $pairs = loadFavorites();
    listRates($pairs);
    exit(0);
}
$cmd = strtolower($argv[1]);
switch ($cmd) {
    case 'list':
        $pairs = loadFavorites();
        listRates($pairs);
        break;
    case 'add':
        if ($argc != 4) {
            echo "Usage: add BASE TARGET\n";
            exit(1);
        }
        $base = strtoupper($argv[2]);
        $target = strtoupper($argv[3]);
        $pairs = loadFavorites();
        $pair = [$base, $target];
        $found = false;
        foreach ($pairs as $p) {
            if ($p[0] == $base && $p[1] == $target) {
                $found = true;
                break;
            }
        }
        if (!$found) {
            $pairs[] = $pair;
            saveFavorites($pairs);
            echo "✅ Added $base/$target\n";
        } else {
            echo "Pair $base/$target already in favorites.\n";
        }
        break;
    case 'remove':
        if ($argc != 4) {
            echo "Usage: remove BASE TARGET\n";
            exit(1);
        }
        $base = strtoupper($argv[2]);
        $target = strtoupper($argv[3]);
        $pairs = loadFavorites();
        $newPairs = [];
        $removed = false;
        foreach ($pairs as $p) {
            if ($p[0] == $base && $p[1] == $target) {
                $removed = true;
            } else {
                $newPairs[] = $p;
            }
        }
        if ($removed) {
            saveFavorites($newPairs);
            echo "✅ Removed $base/$target\n";
        } else {
            echo "Pair $base/$target not found.\n";
        }
        break;
    case 'watch':
        $interval = 60;
        for ($i=2; $i<$argc; $i++) {
            if ($argv[$i] == '--interval' && isset($argv[$i+1])) {
                $interval = (int)$argv[$i+1];
                $i++;
            }
        }
        echo "Watching every {$interval}s. Press Ctrl+C to stop.\n";
        $pairs = loadFavorites();
        while (true) {
            listRates($pairs);
            sleep($interval);
        }
        break;
    default:
        echo "Usage: currency_favorites [list|add BASE TARGET|remove BASE TARGET|watch [--interval N]]\n";
        exit(1);
}
?>

File: currency_favorites.rb
# currency_favorites.rb
#!/usr/bin/env ruby
require 'json'
require 'net/http'
require 'uri'
require 'time'

FAVORITES_FILE = 'favorites.json'
API_URL = 'https://api.frankfurter.app/latest?from='

def load_favorites
  if File.exist?(FAVORITES_FILE)
    data = JSON.parse(File.read(FAVORITES_FILE))
    data['pairs'] || [['USD', 'EUR'], ['USD', 'JPY']]
  else
    [['USD', 'EUR'], ['USD', 'JPY']]
  end
end

def save_favorites(pairs)
  File.write(FAVORITES_FILE, JSON.pretty_generate({ 'pairs' => pairs }))
end

def get_rates(base)
  uri = URI(API_URL + base)
  response = Net::HTTP.get(uri)
  data = JSON.parse(response)
  data['rates']
rescue => e
  puts "Error fetching rates for #{base}: #{e.message}"
  {}
end

def list_rates(pairs)
  if pairs.empty?
    puts "No favorite pairs."
    return
  end
  grouped = {}
  pairs.each do |base, target|
    grouped[base] ||= []
    grouped[base] << target
  end
  puts "\n💱 Favorite Currency Pairs\n"
  grouped.each do |base, targets|
    rates = get_rates(base)
    targets.each do |target|
      if rates[target]
        puts "#{base}/#{target}: #{'%.4f' % rates[target]}"
      else
        puts "#{base}/#{target}: N/A"
      end
    end
  end
  puts "\nUpdated: #{Time.now.strftime('%Y-%m-%d %H:%M:%S')}"
end

cmd = ARGV[0]
case cmd
when nil, 'list'
  pairs = load_favorites
  list_rates(pairs)
when 'add'
  if ARGV.length != 3
    puts "Usage: add BASE TARGET"
    exit 1
  end
  base = ARGV[1].upcase
  target = ARGV[2].upcase
  pairs = load_favorites
  if pairs.include?([base, target])
    puts "Pair #{base}/#{target} already in favorites."
  else
    pairs << [base, target]
    save_favorites(pairs)
    puts "✅ Added #{base}/#{target}"
  end
when 'remove'
  if ARGV.length != 3
    puts "Usage: remove BASE TARGET"
    exit 1
  end
  base = ARGV[1].upcase
  target = ARGV[2].upcase
  pairs = load_favorites
  new_pairs = pairs.reject { |b, t| b == base && t == target }
  if new_pairs.length < pairs.length
    save_favorites(new_pairs)
    puts "✅ Removed #{base}/#{target}"
  else
    puts "Pair #{base}/#{target} not found."
  end
when 'watch'
  interval = 60
  if ARGV.include?('--interval')
    idx = ARGV.index('--interval')
    interval = ARGV[idx+1].to_i if idx && ARGV[idx+1]
  end
  puts "Watching every #{interval}s. Press Ctrl+C to stop."
  pairs = load_favorites
  loop do
    list_rates(pairs)
    sleep(interval)
  end
else
  puts "Usage: currency_favorites [list|add BASE TARGET|remove BASE TARGET|watch [--interval N]]"
end
